ledger/acctonline.go

```diff
@@ -0,0 +1,846 @@
+	ao.acctLookback = cfg.MaxAcctLookback
+}
+
+// loadFromDisk is the 2nd level initialization, and is required before the accountUpdates becomes functional
+// The close function is expected to be call in pair with loadFromDisk
+func (ao *onlineAccounts) loadFromDisk(l ledgerForTracker, lastBalancesRound basics.Round) error {
+	ao.accountsMu.Lock()
+	defer ao.accountsMu.Unlock()
+
+	ao.cachedDBRoundOnline = lastBalancesRound
+	ao.ledger = l
+	err := ao.initializeFromDisk(l, lastBalancesRound)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// initializeFromDisk performs the atomic operation of loading the accounts data information from disk
+// and preparing the accountUpdates for operation.
+func (ao *onlineAccounts) initializeFromDisk(l ledgerForTracker, lastBalancesRound basics.Round) (err error) {
+	ao.dbs = l.trackerDB()
+	ao.log = l.trackerLog()
+
+	err = ao.dbs.Wdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		var err0 error
+		var endRound basics.Round
+		ao.onlineRoundParamsData, endRound, err0 = accountsOnlineRoundParams(tx)
+		if err0 != nil {
+			return err0
+		}
+		if endRound != ao.cachedDBRoundOnline {
+			return fmt.Errorf("last onlineroundparams round %d does not match dbround %d", endRound, ao.cachedDBRoundOnline)
+		}
+
+		onlineAccounts, err0 := onlineAccountsAll(tx, onlineAccountsCacheMaxSize)
+		if err0 != nil {
+			return err0
+		}
+		ao.onlineAccountsCache.init(onlineAccounts, onlineAccountsCacheMaxSize)
+
+		return nil
+	})
+	if err != nil {
+		return
+	}
+
+	ao.accountsq, err = onlineAccountsInitDbQueries(ao.dbs.Rdb.Handle)
+	if err != nil {
+		return
+	}
+
+	ao.deltas = nil
+	ao.accounts = make(map[basics.Address]modifiedOnlineAccount)
+	ao.deltasAccum = []int{0}
+
+	ao.baseOnlineAccounts.init(ao.log, baseAccountsPendingAccountsBufferSize, baseAccountsPendingAccountsWarnThreshold)
+	return
+}
+
+// latest returns the latest round
+func (ao *onlineAccounts) latest() basics.Round {
+	return ao.cachedDBRoundOnline + basics.Round(len(ao.deltas))
+}
+
+// close closes the accountUpdates, waiting for all the child go-routine to complete
+func (ao *onlineAccounts) close() {
+	if ao.accountsq != nil {
+		ao.accountsq.close()
+		ao.accountsq = nil
+	}
+	if ao.voters != nil {
+		ao.voters.close()
+		ao.voters = nil
+	}
+
+	ao.baseOnlineAccounts.prune(0)
+}
+
+// newBlock is the accountUpdates implementation of the ledgerTracker interface. This is the "external" facing function
+// which invokes the internal implementation after taking the lock.
+func (ao *onlineAccounts) newBlock(blk bookkeeping.Block, delta ledgercore.StateDelta) {
+	ao.accountsMu.Lock()
+	ao.newBlockImpl(blk, delta)
+	ao.accountsMu.Unlock()
+	ao.accountsReadCond.Broadcast()
+}
+
+// newBlockImpl is the accountUpdates implementation of the ledgerTracker interface. This is the "internal" facing function
+// which assumes that no lock need to be taken.
+func (ao *onlineAccounts) newBlockImpl(blk bookkeeping.Block, delta ledgercore.StateDelta) {
```

Comment:
nit: when I see `Impl` I usually think of a concrete implementation of an interface, which isn't what this is. A better name would be `newBlockInternal`.

Reply:
It is Java speaks :). We have bunch of functions in ledger with Impl suffix...

Reply:
logged as https://github.com/algorand/go-algorand-internal/issues/2146